### PR TITLE
fix(insights): Refresh stale insight results on shared dashboards

### PR DESCRIPTION
**Pull request by Twixes**
## Problem

We never refreshed the result of a dashboard when it was being viewed in its shared form. This meant that data was stale or non-existent if the results hadn't already been calculated via a logged-in view. [See this customer issue.](https://posthog.slack.com/archives/C0450KBMQBS/p1680515544582769)

## Changes

In #14966 I fixed this in a backend-focused way, but that didn't mesh well with our existing way of refreshing dashboards (https://github.com/PostHog/posthog/pull/14966#discussion_r1170305395 – i.e. we initiate the refresh from the client after initial load, so that there's _something_ to be displayed).

This is an alternative approach which uses the same refreshing logic as already used in the app. The major addition here is that we can now authorize insight retrieval based on a `SharingConfiguration`'s `access_token`, which allows initiating refreshes from shared dashboards client-side.

## How did you test this code?

Pytest.

### CODE REVIEW

**Comment by pauldambra** [comment]
I slightly love this 💖 

There are two new API call failures visiting a shared dashboard when on this branch

| before | after|
|-|--|
|Screenshot 2023-04-20 at 18 47 24|Screenshot 2023-04-20 at 18 46 13|

Other than that as far I can tell this is spot on...

There are two things that occur to me

1. We really don't get any analytics from shared dashboards. Would be cool to send ourselves an event when we're serving one. So we can see if they're being used. Which could be useful as a caching signal too in the future. This feels like a follow-up more than essential though

2. The "real risk" here is it's a new route to hit CH. So, someone gets to the top spot on HN with a page that has a shared dashboard embedded. Or someone decides to take the API call and replay it to DDOS us. We have rate limiting now so there's some protection here. But ideally, the rate limits would be lower for a shared dashboard API call than a "normal" API call.

**Comment by Twixes** [comment]
Thanks for the review @pauldambra! The API failures should be resolved now.
As for analytics, that's a good point, but I think for another PR.
As for CH load, let's get https://github.com/PostHog/posthog/pull/14966 in before this to reduce the risk.
Does that sounds good?

**Review by pauldambra — APPROVED** [review]
I think this is a really cool improvement to shared dashboards! 😍

File: posthog/models/sharing_configuration.py [hunk 1]
```diff
@@ -19,3 +20,15 @@ class SharingConfiguration(models.Model):
...
         max_length=400, null=True, blank=True, default=get_default_access_token, unique=True
     )
+
+    def get_connected_insight_ids(self) -> List[int]:
+        if self.insight:
+            return [self.insight.id]
+        elif self.dashboard:
+            # Check whether this sharing configuration's dashboard contains this insight
+            return list(
+                self.dashboard.tiles.exclude(deleted=True, insight__isnull=True, insight__deleted=True).values_list(
```
Comment (mariusandra) [inline posthog/models/sharing_configuration.py:30]: Do you want to exclude `deleted=True AND insight__isnull=True AND insight__deleted=True` or the same with `OR`-s? [According to Django](https://docs.djangoproject.com/en/4.2/ref/models/querysets/#exclude) it'll do the `AND`-s, though it looks like we'd want to exclude insights when any of these is true?

File: frontend/__snapshots__/components-sharing--dashboard-sharing.png
Comment (mariusandra) [inline frontend/__snapshots__/components-sharing--dashboard-sharing.png:1]: Seems to be an actual visual regression somehow?

**Review by mariusandra — COMMENTED** [review]
Found some small things to comment on, otherwise looks good!
Yet my biggest worry still remains: getting all of Reddit to refresh a dashboard all at once might send us sudden bursts of death. Some kind of quick "418 busy now, ask again later" response from the API with a frontend that knows to "try again in 2, 5, 10, 20, ... sec" might be better.

Or, async django.

File: frontend/__snapshots__/components-sharing--dashboard-sharing.png
Comment (Twixes) [reply frontend/__snapshots__/components-sharing--dashboard-sharing.png:1]: Small flyby improvement, this just centers the lock icon

File: posthog/models/sharing_configuration.py (hunk 1 above)
Comment (Twixes) [reply posthog/models/sharing_configuration.py]: Good catch! Now, I just added tests to make sure that insight with any link in the chain deleted (insight, tile, dashboard) cannot be retrieved with the sharing access token _after fixing this_, but it turns out this already is guaranteed by all the implicit `objects` vs. `objects_including_soft_deleted` filtering that @pauldambra added – so we're safe. (Nevertheless added a few test cases + cleaned up some handling of `.deleted` model instances.)

**Comment by Twixes** [comment]
If that radical scenario happens: 1. we'll be happy we went viral in some way I guess, I don't think that's happened before; 2. after #14966 we'd just have a momentary problem with handling ` web` pod traffic (if the load is really extreme I suppose), everything else should be safe.

**Pull request merged**